src/v/archival/probe.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "archival/types.h"
+#include "model/fundamental.h"
+#include "seastarx.h"
+
+#include <seastar/core/metrics_registration.hh>
+
+#include <cstdint>
+
+namespace archival {
+
+/// \brief Per-ntp archval service probe
+///
+/// For every NTP we need to track how much log we've already uploaded,
+/// how much data we're missing (e.g. because it was compacted before upload)
+/// and how much work remains.
+///
+/// The unit of measure is offset delta.
+class ntp_level_probe {
+public:
+    ntp_level_probe(per_ntp_metrics_disabled disabled, model::ntp ntp);
+
+    /// Register log-segment upload
+    void uploaded(model::offset offset_delta) { _uploaded += offset_delta; }
+
+    /// Register gap
+    void gap_detected(model::offset offset_delta) { _missing += offset_delta; }
+
+    /// Register the offset the ought to be uploaded
+    void upload_lag(model::offset offset_delta) { _pending = offset_delta; }
+
+private:
+    /// NTP of the probe
+    model::ntp _ntp;
```

Comment:
not related to this tix, but we store so many copies of `model::ntp` in the tree that we should really optimize these away from a string into 3 integers.